[CDAP-18586] Deprecate getApplicationSpecification in ProgramStatusTriggerInfo

wyzhang opened the pull request:
Why:
    ApplicationSpecification can be fairly large due to it storing
    app config and plugin info. ProgramStatusTriggerInfo doesn't
    need entire application spec, only the application name for
    logging purpose.

    A large ProgramStatusTriggerInfo can be problematic because
    it is included in run record. Large run record may lead to
    out of memory issues.

    What:
    This PR does
    - deprecate getApplicationSpecification() API
    - introduce getApplicationName() API

    Once the API gets removed, we can replace ApplicationSpecification
    with Application name, so reduce the size of trigger info

wyzhang commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/trigger/DefaultProgramStatusTriggerInfo.java:
```diff
@@ -142,7 +140,7 @@ public void writeExternal(ObjectOutput out) throws IOException {
   public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
     namespace = in.readUTF();
-    applicationSpecification = GSON.fromJson((String) in.readObject(), ApplicationSpecification.class);
+    applicationName = in.readUTF();
```
Tried to maintain backward compatibility, but seems impossible? As a result, this change is backward-incompatible. 
Let me know you see a backward compatible way. 

Also cannot tell if we ever serialize this?

wyzhang replied:
See comments above. 

Short summary is: 
- Serialization happens in spark, but appears to be mostly transient (i.e.. serialized for wire transferring) 
- Seems no way to maintain backward compatible, but might be ok given above.

sagarkapare commented on cdap-api/src/main/java/io/cdap/cdap/api/schedule/ProgramStatusTriggerInfo.java:
```diff
@@ -36,9 +36,9 @@
+   * @return The name of the application that contains the triggering program.
    */
-  ApplicationSpecification getApplicationSpecification();
+  String getApplicationName();
```
I think we first need to deprecate the `getApplicationSpecification` API before removing it since its part of cdap-api package.

wyzhang replied:
Will do.

wyzhang replied:
Done

sagarkapare commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/trigger/DefaultProgramStatusTriggerInfo.java:
```diff
@@ -142,7 +140,7 @@ public void writeExternal(ObjectOutput out) throws IOException {
   public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
     namespace = in.readUTF();
-    applicationSpecification = GSON.fromJson((String) in.readObject(), ApplicationSpecification.class);
+    applicationName = in.readUTF();
```
It will fail when it reads the old serialized version of the `ProgramStatusTriggerInfo` which had application specification in it, right?

wyzhang replied:
Correct. This change is not backward compatible, but 
1. Could you point me to where are serializing it and store externally ? Did some digging, but don't fine any obvious place. 
2.  Given the way we do seralization here, not sure if I see a backward-compatible way. Any thoughts ?

sagarkapare replied:
I think this is stored in the schedule store, when `PUT /v3/namespaces/<namespace-id>/apps/<app-id>/versions/<version-id>/schedules/<schedule-id>` API is invoked. ProgramStatusTriggerInfo will be present for triggers of type program status. Details here https://cdap.atlassian.net/wiki/spaces/DOCS/pages/477560983/Lifecycle+Microservices. The API handler is here https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java#L682

TriggerInfo is deserialized and passed to the Workflow run which is used by Datapipeline workflow - https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java#L383

sagarkapare replied:
From backward compatibility perspective I think something like following might work :) 

1. Have a new class which will have `serialVersionUID` (to allow us to change any future modification to the class members) and data format version (for future ser-de purpose).
```java
public class VersionedProgramStatusTriggerInfo implements Externalizable {
static final long serialVersionUID = -123456789L;
private int DATA_FORMAT_VERSION=1;
private final String namespace;
private final String appName; // not the application specification from the existing one
... // other members

// constructor called after deserializing the old formatted data from readResolve method of DefaultProgramStatusTriggerInfo
public VersionedProgramStatusTriggerInfo(String namespace, String applicationName, ...) {
   this.namespace = namespace;
   this.applicationName = applicationName;
}

   @Override
    public void writeExternal(ObjectOutput out) throws IOException {
       out.writeInt(DATA_FORMAT_VERSION);
       // write other members;
    }

    @Override
    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
      // read members per expectation of this class.
     // this will always work since java will de-serialize to this class the content written by writeExternal of this class
    }
}
```


2. Add `readResolve` method to the existing `DefaultProgramStatusTriggerInfo`, which instructs which object to return on deserialization of the old data.

```java
public class DefaultProgramStatusTriggerInfo extends AbstractTriggerInfo
  implements ProgramStatusTriggerInfo, Externalizable {

// No changes to member variables of this class

// Add new private method readResolve which returns instance of `VersionedProgramStatusTriggerInfo`. `readResolve` is called after readExternal of this class is complete so should have access to all member variables including applicationSpecification from which app name can be derived.
    private Object readResolve() {
      return new VersionedProgramStatusTriggerInfo(namespace, applicationSpecification.getAppName(), ...);
    }
}
```

sagarkapare replied:
Taking more closer look, it does not look like schedule store stores the application specification itself. It only stores the [ProgramId and status](https://github.com/cdapio/cdap/blob/develop/cdap-proto/src/main/java/io/cdap/cdap/proto/ProtoTrigger.java#L239) .

Schedule info is passed from app fabric to program container as [system argument](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java#L90-L91) in the serialized form. Application specifications are [fetched on the fly from app meta store](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/trigger/TriggerInfoContext.java#L71) during this serialization. On program container side, this information is made available to the programs through context after deserialization of [system arguments](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java#L253-L257).

So it looks like changing `DefaultProgramStatusTriggerInfo`'s  writeExternal and readExternal methods to write and read application name instead of application specification object does NOT impact compatibility as its only done for wire transfer and not actually stored in any persistent storage. 

So looks like we need to 
1. Deprecate ProgramStatusTriggerInfo.getApplicationSpecification method.
2. Add new ProgramStatusTriggerInfo.getApplication API.
3. Keep DefaultProgramStatusTriggerInfo.readExternal and DefaultProgramStatusTriggerInfo.writeExternal methods as it is while adding implementation of DefaultProgramStatusTriggerInfo.getApplication() method.
4. ApplicationSpecification will still be part of run record at this point since its part of system argument.
5. When we remove `ProgramStatusTriggerInfo.getApplicationSpecification` actually at that point, run record size will decrease.

wyzhang replied:
For serialVersionUID, 
- Sadly the original code should have specified one from day one instead of letting java pick a default, which makes it impossible to change any field. 
- From a quick look, serialization  happens only in spark, which is mostly for transient purpose, but there theoretically there is risk in spark checkpoint I suppose. If spark checkpoint is less of a concern, we can introduce a custom serial version uid when changing the field.

For deprecating getAppliationSpecification 
- Done, so we can prepare for getting rid of app spec in next release.

wyzhang replied:
For readResolve(), 
It gets called after object being created from byte stream but before being returned. This doesn't solve the backward compatibility issue where new code tries to read seralized object by old code. As a matter of fact, AFAIK, it might be impossible to maintain backward compatibility  in this case. 

Any change to this class will result in InvalidClassException if we attempt to deserialize an object serialized by old code.

sagarkapare replied:
readResolve will solve the backward compatibility problem. We need to keep the `DefaultProgramStatusTriggerInfo` which we cant change. So instance of `DefaultProgramStatusTriggerInfo` will be created on de-serialization and this will work since the stored data will have appspec right? Now from that point we can return the `VersionedProgramStatusTriggerInfo` instance from ``DefaultProgramStatusTriggerInfo.readResolve` 

Since readResolve is private method it wont impact java generated serial version uid.

wyzhang replied:
If we have a serialized DefaultProgramStatusTriggerInfo object stored as byte array somewhere, new code will attempt to deserialize it into VersionedProgramStatusTriggerInfo, which will fail, doesn't it?

wyzhang replied:
Don't think java serialization tracks which class the serialized bytes correspond to, thus suspecting if the approach above would help

wyzhang committed:
[CDAP-18586] Deprecate getApplicationSpecification in ProgramStatusTriggerInfo
             in preparation of removing ApplicationSpecification from
             DefaultProgramStatusTriggerInfo

Why:
ApplicationSpecification can be fairly large due to it storing
app config and plugin info. ProgramStatusTriggerInfo doesn't
need entire application spec, only the application name for
logging purpose.

A large ProgramStatusTriggerInfo can be problematic because
it is included in run record. Large run record may lead to
out of memory issues.

What:
This PR does
- deprecate getApplicationSpecification() API
- introduce getApplicationName() API

Once the API gets removed, we can replace ApplicationSpecification
with Application name, so reduce the size of trigger info

wyzhang commented:
Ready for review again. PTAL
